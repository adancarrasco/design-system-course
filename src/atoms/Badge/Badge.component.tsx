import React from 'react'
import styled from '@emotion/styled'

export interface IBadgeProps extends BadgeStyledProps {
  userName?: string
  initials?: string
}

export const getInitials = (userName) => {
  if (!userName) return null
  const [firstName, lastName] = userName.split(' ')
  return firstName[0] + lastName[0]
}

const Badge: React.FC<IBadgeProps> = ({ userName, backgroundImage, ...rest }) => {
  const initials = getInitials(userName)
  const BadgeComponent =
    backgroundImage || !initials ? (
      <BadgeStyled backgroundImage={backgroundImage} {...rest} />
    ) : (
      <BadgeStyled {...rest}>{initials}</BadgeStyled>
    )
  return BadgeComponent
}

interface BadgeStyledProps {
  backgroundImage?: string
  size?: string
  backgroundColor?: string
}

const defaultSize = '60px'

const BadgeStyled = styled.span<BadgeStyledProps>`
  width: ${(props) => props.size ?? defaultSize};
  height: ${(props) => props.size ?? defaultSize};
  background-image: ${(props) => 'url("' + props.backgroundImage + '")' ?? ''};
  background-color: ${(props) => props.backgroundColor ?? '#ababab;'};
  background-size: ${(props) => props.size ?? defaultSize};
  display: inline-flex;
  border-radius: 50%;
  justify-content: center;
  flex-direction: column;
  text-align: center;
`

export default Badge
