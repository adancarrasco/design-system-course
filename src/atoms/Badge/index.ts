export { default, getInitials } from './Badge.component'
export type { IBadgeProps } from './Badge.component'
