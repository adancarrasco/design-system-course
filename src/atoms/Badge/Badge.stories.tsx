import React from 'react'
import { storiesOf } from '@storybook/react'

import Badge from '.'

storiesOf('Atoms|Badge', module)
  .add('Default', () => <Badge />)
  .add('Background image', () => (
    <Badge backgroundImage="https://avatars2.githubusercontent.com/u/3975603?s=80&v=4" />
  ))
  .add('Initials', () => <Badge userName="AdÃ¡n Carrasco" />)
