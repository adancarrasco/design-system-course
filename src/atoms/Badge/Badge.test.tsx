import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Badge, { getInitials, IBadgeProps } from '.'

describe('Badge Component', () => {
  const renderComponent = ({ userName, backgroundImage }: IBadgeProps) =>
    render(<Badge userName={userName} backgroundImage={backgroundImage} />)

  it('matches snapshot', () => {
    const props = {
      userName: 'Adan Carrasco',
    }
    const { getByText } = renderComponent(props)
    const container = getByText(getInitials(props.userName))
    expect(container).toMatchSnapshot()
  })

  it('should render', () => {
    const props = {
      userName: 'Adan Carrasco',
    }
    const { getByText } = renderComponent(props)
    const container = getByText(getInitials(props.userName))
    expect(container).toBeInTheDocument()
  })

  it('should return initials', () => {
    const initials = getInitials('Adan Carrasco')
    expect(initials).toBe('AC')
  })
})
