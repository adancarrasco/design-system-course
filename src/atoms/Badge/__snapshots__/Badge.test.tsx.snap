// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge Component matches snapshot 1`] = `
.emotion-0 {
  width: 60px;
  height: 60px;
  background-image: url("undefined");
  background-color: #ababab;
  background-size: 60px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 50%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  text-align: center;
}

<span
  class="emotion-0"
>
  AC
</span>
`;
